File: SquirrelConfig.m
#import "SquirrelConfig.h"

#import <rime_api.h>

@implementation SquirrelConfig {
  NSMutableDictionary* _cache;
  RimeConfig _config;
  SquirrelConfig* _baseConfig;
}

- (instancetype)init {
  if (self = [super init]) {
    _cache = [[NSMutableDictionary alloc] init];
    _colorSpace = @"srgb";
  }
  return self;
}

- (BOOL)openBaseConfig {
  [self close];
  _isOpen = !!rime_get_api()->config_open("squirrel", &_config);
  return _isOpen;
}

- (BOOL)openWithSchemaId:(NSString*)schemaId
              baseConfig:(SquirrelConfig*)baseConfig {
  [self close];
  _isOpen = !!rime_get_api()->schema_open(schemaId.UTF8String, &_config);
  if (_isOpen) {
    _schemaId = schemaId;
    _baseConfig = baseConfig;
  }
  return _isOpen;
}

- (void)close {
  if (_isOpen) {
    rime_get_api()->config_close(&_config);
    _baseConfig = nil;
    _isOpen = NO;
  }
}

- (void)dealloc {
  [self close];
}

- (BOOL)hasSection:(NSString*)section {
  if (_isOpen) {
    RimeConfigIterator iterator = {0};
    if (rime_get_api()->config_begin_map(&iterator, &_config,
                                         section.UTF8String)) {
      rime_get_api()->config_end(&iterator);
      return YES;
    }
  }
  return NO;
}

- (BOOL)getBool:(NSString*)option {
  return [self getOptionalBool:option].boolValue;
}

- (int)getInt:(NSString*)option {
  return [self getOptionalInt:option].intValue;
}

- (double)getDouble:(NSString*)option {
  return [self getOptionalDouble:option].doubleValue;
}

- (NSNumber*)getOptionalBool:(NSString*)option {
  NSNumber* cachedValue = [self cachedValueOfObjCType:@encode(BOOL)
                                               forKey:option];
  if (cachedValue) {
    return cachedValue;
  }
  Bool value;
  if (_isOpen &&
      rime_get_api()->config_get_bool(&_config, option.UTF8String, &value)) {
    return _cache[option] = @(!!value);
  }
  return [_baseConfig getOptionalBool:option];
}

- (NSNumber*)getOptionalInt:(NSString*)option {
  NSNumber* cachedValue = [self cachedValueOfObjCType:@encode(int)
                                               forKey:option];
  if (cachedValue) {
    return cachedValue;
  }
  int value;
  if (_isOpen &&
      rime_get_api()->config_get_int(&_config, option.UTF8String, &value)) {
    return _cache[option] = @(value);
  }
  return [_baseConfig getOptionalInt:option];
}

- (NSNumber*)getOptionalDouble:(NSString*)option {
  NSNumber* cachedValue = [self cachedValueOfObjCType:@encode(double)
                                               forKey:option];
  if (cachedValue) {
    return cachedValue;
  }
  double value;
  if (_isOpen &&
      rime_get_api()->config_get_double(&_config, option.UTF8String, &value)) {
    return _cache[option] = @(value);
  }
  return [_baseConfig getOptionalDouble:option];
}

- (NSString*)getString:(NSString*)option {
  NSString* cachedValue = [self cachedValueOfClass:[NSString class]
                                            forKey:option];
  if (cachedValue) {
    return cachedValue;
  }
  const char* value =
      _isOpen ? rime_get_api()->config_get_cstring(&_config, option.UTF8String)
              : NULL;
  if (value) {
    return _cache[option] = @(value);
  }
  return [_baseConfig getString:option];
}

- (NSColor*)getColor:(NSString*)option {
  NSColor* cachedValue = [self cachedValueOfClass:[NSColor class]
                                           forKey:option];
  if (cachedValue) {
    return cachedValue;
  }
  NSColor* color = [self colorFromString:[self getString:option]];
  if (color) {
    _cache[option] = color;
    return color;
  }
  return [_baseConfig getColor:option];
}

- (SquirrelAppOptions*)getAppOptions:(NSString*)appName {
  NSString* rootKey = [@"app_options/" stringByAppendingString:appName];
  SquirrelMutableAppOptions* appOptions =
      [[SquirrelMutableAppOptions alloc] init];
  RimeConfigIterator iterator;
  rime_get_api()->config_begin_map(&iterator, &_config, rootKey.UTF8String);
  while (rime_get_api()->config_next(&iterator)) {
    // NSLog(@"DEBUG option[%d]: %s (%s)", iterator.index, iterator.key,
    // iterator.path);
    BOOL value = [self getBool:@(iterator.path)];
    appOptions[@(iterator.key)] = @(value);
  }
  rime_get_api()->config_end(&iterator);
  return [appOptions copy];
}

#pragma mark - Private methods

- (id)cachedValueOfClass:(Class)aClass forKey:(NSString*)key {
  id value = [_cache objectForKey:key];
  if ([value isMemberOfClass:aClass]) {
    return value;
  }
  return nil;
}

- (NSNumber*)cachedValueOfObjCType:(const char*)type forKey:(NSString*)key {
  id value = [_cache objectForKey:key];
  if ([value isMemberOfClass:NSNumber.class] &&
      !strcmp([value objCType], type)) {
    return value;
  }
  return nil;
}

- (NSColor*)colorFromString:(NSString*)string {
  if (string == nil) {
    return nil;
  }

  int r = 0, g = 0, b = 0, a = 0xff;
  if (string.length == 10) {
    // 0xffccbbaa
    sscanf(string.UTF8String, "0x%02x%02x%02x%02x", &a, &b, &g, &r);
  } else if (string.length == 8) {
    // 0xccbbaa
    sscanf(string.UTF8String, "0x%02x%02x%02x", &b, &g, &r);
  }
  if ([self.colorSpace isEqualToString:@"display_p3"]) {
    return [NSColor colorWithDisplayP3Red:r / 255.0
                                    green:g / 255.0
                                     blue:b / 255.0
                                    alpha:a / 255.0];
  } else {  // sRGB by default
    return [NSColor colorWithSRGBRed:r / 255.0
                               green:g / 255.0
                                blue:b / 255.0
                               alpha:a / 255.0];
  }
}

@end

File: SquirrelConfig.swift
//
//  SquirrelConfig.swift
//  Squirrel
//
//  Created by Leo Liu on 5/9/24.
//

import AppKit

class SquirrelConfig {  
  private let rimeAPI = rime_get_api().pointee
  private(set) var isOpen = false
  var schemaID: String = ""
  
  private var cache: Dictionary<String, Any> = [:]
  private var config: RimeConfig = .init()
  private var baseConfig: SquirrelConfig?
  
  func openBaseConfig() -> Bool {
    close()
    isOpen = rimeAPI.config_open("squirrel", &config)
    return isOpen
  }
  
  func open(schemaID: String, baseConfig: SquirrelConfig?) -> Bool {
    close()
    isOpen = rimeAPI.schema_open(schemaID, &config)
    if isOpen {
      self.schemaID = schemaID
      self.baseConfig = baseConfig
    }
    return isOpen
  }
  
  func close() {
    if isOpen {
      let _ = rimeAPI.config_close(&config)
      baseConfig = nil
      isOpen = false
    }
  }
  
  deinit {
    close()
  }
  
  func has(section: String) -> Bool {
    if isOpen {
      var iterator: RimeConfigIterator = .init()
      if rimeAPI.config_begin_map(&iterator, &config, section) {
        rimeAPI.config_end(&iterator)
        return true
      }
    }
    return false
  }
  
  func getBool(_ option: String) -> Bool? {
    if let cachedValue = cachedValue(of: Bool.self, forKey: option) {
      return cachedValue
    }
    var value = false
    if isOpen && rimeAPI.config_get_bool(&config, option, &value) {
      cache[option] = value
      return value
    }
    return baseConfig?.getBool(option)
  }
  
  func getInt(_ option: String) -> Int? {
    if let cachedValue = cachedValue(of: Int.self, forKey: option) {
      return cachedValue
    }
    var value: Int32 = 0
    if isOpen && rimeAPI.config_get_int(&config, option, &value) {
      cache[option] = value
      return Int(value)
    }
    return baseConfig?.getInt(option)
  }
  
  func getDouble(_ option: String) -> Double? {
    if let cachedValue = cachedValue(of: Double.self, forKey: option) {
      return cachedValue
    }
    var value: Double = 0
    if isOpen && rimeAPI.config_get_double(&config, option, &value) {
      cache[option] = value
      return value
    }
    return baseConfig?.getDouble(option)
  }
  
  func getString(_ option: String) -> String? {
    if let cachedValue = cachedValue(of: String.self, forKey: option) {
      return cachedValue
    }
    if isOpen, let value = rimeAPI.config_get_cstring(&config, option) {
      cache[option] = String(cString: value)
      return String(cString: value)
    }
    return baseConfig?.getString(option)
  }
  
  func getColor(_ option: String, inSpace colorSpace: SquirrelTheme.RimeColorSpace) -> NSColor? {
    if let cachedValue = cachedValue(of: NSColor.self, forKey: option) {
      return cachedValue
    }
    if let colorStr = getString(option), let color = color(from: colorStr, inSpace: colorSpace) {
      cache[option] = color
      return color
    }
    return baseConfig?.getColor(option, inSpace: colorSpace)
  }
  
  func getAppOptions(_ appName: String) -> Dictionary<String, Bool> {
    let rootKey = "app_options/\(appName)"
    var appOptions = [String : Bool]()
    var iterator = RimeConfigIterator()
    let _ = rimeAPI.config_begin_map(&iterator, &config, rootKey)
    while rimeAPI.config_next(&iterator) {
      // print("[DEBUG] option[\(iterator.index)]: \(String(cString: iterator.key)), path: (\(String(cString: iterator.path))")
      if let key = iterator.key, let path = iterator.path, let value = getBool(String(cString: path)) {
        appOptions[String(cString: key)] = value
      }
    }
    rimeAPI.config_end(&iterator)
    return appOptions
  }
  
  // isLinear
  func updateCandidateListLayout(prefix: String) -> Bool {
    let candidateListLayout = getString("\(prefix)/candidate_list_layout")
    if candidateListLayout == "stacked" {
      return false
    } else if candidateListLayout == "linear" {
      return true
    } else {
      // Deprecated. Not to be confused with text_orientation: horizontal
      return getBool("\(prefix)/horizontal") ?? false
    }
  }
  
  // isVertical
  func updateTextOrientation(prefix: String) -> Bool {
  let textOrientation = getString("\(prefix)/text_orientation")
    if textOrientation == "horizontal" {
      return false
    } else if textOrientation == "vertical" {
      return true
    } else {
      // Deprecated.
      return getBool("\(prefix)/vertical") ?? true
    }
  }
}

private extension SquirrelConfig {
  func cachedValue<T>(of: T.Type, forKey key: String) -> T? {
    return cache[key] as? T
  }
  
  func color(from colorStr: String, inSpace colorSpace: SquirrelTheme.RimeColorSpace) -> NSColor? {
    if let matched = try? /0x([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})/.wholeMatch(in: colorStr) {
      let (_, a, b, g, r) = matched.output
      return color(alpha: Int(a, radix: 16)!, red: Int(r, radix: 16)!, green: Int(g, radix: 16)!, blue: Int(b, radix: 16)!, colorSpace: colorSpace)
    } else if let matched = try? /0x([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})/.wholeMatch(in: colorStr) {
      let (_, b, g, r) = matched.output
      return color(alpha: 255, red: Int(r, radix: 16)!, green: Int(g, radix: 16)!, blue: Int(b, radix: 16)!, colorSpace: colorSpace)
    } else {
      return nil
    }
  }
  
  func color(alpha: Int, red: Int, green: Int, blue: Int, colorSpace: SquirrelTheme.RimeColorSpace) -> NSColor {
    switch colorSpace {
    case .displayP3:
      return NSColor(displayP3Red: CGFloat(red) / 255,
                     green: CGFloat(green) / 255,
                     blue: CGFloat(blue) / 255,
                     alpha: CGFloat(alpha) / 255)
    case .sRGB:
      return NSColor(srgbRed: CGFloat(red) / 255,
                     green: CGFloat(green) / 255,
                     blue: CGFloat(blue) / 255,
                     alpha: CGFloat(alpha) / 255)
    }
  }
}
